### PR TITLE
sql: support scalar expressions (& placeholders) for alter range ... relocate

### LINKED ISSUE #73315: kv: add placeholders to ALTER RANGE RELOCATE
#72305 introduced a new ALTER RANGE RELOCATE syntax to move leases and replicas around. The syntax does not support placeholder values for the range/store ids. This issue is to track to extending the syntax to support placeholders. 

cc: @knz

### PR DESCRIPTION
**Pull request by knz**
Fixes #73315.
See the individual commits for details.

### CODE REVIEW

**Comment by knz** [comment]
@rafiss @otan do you have some yacc-fu that'd be useful here? By defining clauses as `FROM a_expr TO a_expr`,  we get a shift/reduce conflict on TO inside the interval syntax (e.g. INTERVAL DAYS TO MONTHS).  Is there a way to ask yacc to prefer shifting in that case?

**Comment by otan** [comment]
you could try use `%prec` or maybe even consider b_expr or below (or force parenthesis around FROM/TO)

**Comment by knz** [comment]
> you could try use %prec

can you explain how that works? Or maybe link to a doc that explains?

> or maybe even consider b_expr or below

unfortunately the interval syntax is in d_expr (the narrowest one)

> or force parenthesis around FROM/TO

yeah I thought about that, and if there's no solution via %prec, I'll put the choice to a vote on the KV team: either `TO y FROM x`, or `FROM (x) TO (y)`

**Comment by otan** [comment]
Best I can find is https://www.ibm.com/docs/en/zos/2.3.0?topic=section-precedence-in-grammar-rules, I only have a vague understanding of it cargo culting from pg

**Review by RaduBerinde — COMMENTED** [review]
:lgtm: modulo the syntax discussion

If we support placeholders now, it would be good to have some tests for that. I can't find many existing tests for RELOCATE, maybe we can add some similar to `TestSplitAt`

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/73807#-:-MquSb5I04z28wTJInP8:bdb1okh)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @arulajmani and @lunevalex)

**Review by lunevalex — APPROVED** [review]
LGTM

**Comment by knz** [comment]
Thanks to the goyacc source code, I found out that it uses a super old algorithm that only supports _real_ token names in `%prec` (not pseudo-tokens) for the shifting rule. Once I understood that, I was able to fix the syntax.

**Comment by knz** [comment]
For future reference, I troubleshooted my yacc problem with the following mini-grammar:

```yacc
%token RELOCATE FROM TO FOR SELECT DAY TO MINUTE

%left INTERVAL_SHORT
%left TO

%%

top:
   RELOCATE FROM a TO a FOR SELECT

a: b

b: DAY           %prec INTERVAL_SHORT
 | DAY TO MINUTE %prec TO
```

File: pkg/ccl/logictestccl/testdata/logic_test/tenant_unsupported [hunk 1]
```diff
@@ -73,10 +73,10 @@ statement error operation is unsupported in multi-tenancy mode
 ALTER RANGE RELOCATE LEASE TO 2 FOR SELECT range_id from crdb_internal.ranges where table_name = 'kv'
 
 statement error operation is unsupported in multi-tenancy mode
-ALTER RANGE 1 RELOCATE FROM 1 TO 2
+ALTER RANGE 1 RELOCATE TO 2 FROM 1
```
Comment (otan) [inline pkg/ccl/logictestccl/testdata/logic_test/tenant_unsupported:76]: revert?

Comment (knz) [reply pkg/ccl/logictestccl/testdata/logic_test/tenant_unsupported:76]: Good call. Done.

**Comment by knz** [comment]
I have extended the PR with a 2nd commit that also adds placeholder support in the first position, so that this PR closes the linked issue entirely.

Please take another look.

**Review by otan — APPROVED** [review]
lgtm, some small test failure that look like simple rewrites

**Comment by knz** [comment]
TFYRs!

bors r=lunevalex,otan

**Comment by knz** [comment]
bors r=lunevalex,otan

**Comment by knz** [comment]
bors r=lunevalex,otan

**Comment by RaduBerinde** [comment]
Opt changes LGTM.

**Pull request merged**